docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,477 @@
+  - `spec.worker.deployments[i].class` field must be unique within a ClusterClass.
+- For object updates:
+  - `spec.worker.deployments` supports adding new deployment classes.
+
+##### Cluster
+- For object creation:
+  - `spec.topology` and `spec.controlPlaneRef` cannot be simultaneously set.
+  - If `spec.topology.class` is set, `spec.topology.version` cannot be empty and must be a valid semver.
+
+- For object updates:
+  - If spec.topology.class is in use, it cannot be unset or modified.
+  - spec.topology.version cannot be unset and must be a valid semver, if being updated.
+  - `spec.topology.version` cannot be downgraded.
+  - `spec.topology.controlPlane.replicas` can be updated.
+  - `spec.topology.worker.machineDeployments[i].replicas` can be updated.
+  - `spec.topology.worker.machineDeployments[i].name` field must be unique within a Cluster
+  - A set of worker nodes can be added to or removed from the `spec.topology.worker.machineDeployments` list.
+
+#### Behaviors
+This section lists out the behavior for Cluster objects using `ClusterClass` in case of creates and updates.
+
+##### Create a new Cluster using ClusterClass object
+1. User creates a ClusterClass object.
+   ```yaml
+    apiVersion: cluster.x-k8s.io/v1alpha4
+    kind: ClusterClass
+    metadata:
+      name: mixed
+      namespace: bar
+    spec:
+      controlPlane:
+        ref:
+          apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
+          kind: KubeadmControlPlaneTemplate
+          name: vsphere-prod-cluster-template-kcp
+      worker:
+        deployments:
+        - class: linux-worker
+          template:
+            bootstrap:
+              ref:
+                apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
+                kind: KubeadmConfigTemplate
+                name: existing-boot-ref
+            infrastructure:
+              ref:
+                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+                kind: VSphereMachineTemplate
+                name: linux-vsphere-template
+        - class: windows-worker
+          template:
+            bootstrap:
+              ref:
+                apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
+                kind: KubeadmConfigTemplate
+                name: existing-boot-ref-windows
+            infrastructure:
+              ref:
+                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+                kind: VSphereMachineTemplate
+                name: windows-vsphere-template
+      infrastructure:
+        ref:
+          apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+          kind: VSphereClusterTemplate
+          name: vsphere-prod-cluster-template
+   ```
+2. User creates a cluster using the class name and defining the managed topology.
+   ```yaml
+    apiVersion: cluster.x-k8s.io/v1alpha4
+    kind: Cluster
+    metadata:
+      name: foo
+      namespace: bar
+    spec:
+      topology:
+        class: mixed
+        version: v1.19.1
```

Comment:
Should we consider as part of the upgrade strategy talks on the ClusterClass side to allow folks to specify a minimum version that would then be rolled out to all clusters?

Reply:
Is this necessary for adopting existing clusters (which do not use a CC), since the `version` field is mandatory for creating new clusters with CC.